File: code_samples/typescript/design_patterns_observer.ts
/**
 * Design Patterns: Observer
 * AI/ML Training Sample
 */

interface IObserver {
    data: string;
    process(input: string): void;
    validate(): boolean;
}

class Observer implements IObserver {
    data: string;
    
    constructor() {
        this.data = "";
    }
    
    process(input: string): void {
        this.data = input;
    }
    
    getData(): string {
        return this.data;
    }
    
    validate(): boolean {
        return this.data.length > 0;
    }
}

// Example usage
const instance = new Observer();
instance.process("example");
console.log(`Data: ${instance.getData()}`);
console.log(`Valid: ${instance.validate()}`);

export { Observer, IObserver };

File: code_samples/elixir/design_patterns_observer.ex
# Design Patterns: Observer
# AI/ML Training Sample

defmodule Observer do
  defstruct data: ""
  
  def new(), do: %Observer{}
  
  def process(%Observer{} = struct, input) do
    %{struct | data: input}
  end
  
  def get_data(%Observer{data: data}), do: data
  
  def validate(%Observer{data: data}) do
    String.length(data) > 0
  end
end

# Example usage
instance = Observer.new()
updated = Observer.process(instance, "example")
IO.puts("Data: " <> Observer.get_data(updated))
IO.puts("Valid: " <> to_string(Observer.validate(updated)))
